File: packages/integrations/autogen/cred_autogen/tool.py
"""Cred AutoGen integration.

Returns FunctionTools that AutoGen agents can call for delegated OAuth access.
The tool schema matches the Cred MCP tool spec:

    name:   cred_delegate
    params: { service: str, scopes: list[str] }

user_id and app_client_id are pre-configured at factory time and are not
agent-controlled inputs.
"""

from __future__ import annotations

import json
from typing import Annotated, Any, Optional

from autogen_core.tools import FunctionTool

from cred import Cred


def cred_delegate_tool(
    agent_token: str,
    user_id: str,
    app_client_id: str,
    base_url: str,
    token_format: str = "raw",
) -> FunctionTool:
    """Create an AutoGen FunctionTool for credential delegation.

    The returned tool is pre-configured with the user's identity and app context.
    At runtime, the agent supplies ``service`` and ``scopes``.

    Args:
        agent_token:    Agent token issued by Cred (starts with ``cred_at_``).
        user_id:        The user whose credentials to delegate.
        app_client_id:  The Cred app client ID to delegate for.
        base_url:       Override the API base URL.

    Returns:
        A ``FunctionTool`` ready to use with AutoGen agents.

    Raises:
        ConsentRequiredError: (from the tool at runtime) when the user hasn't
            connected the requested service. The error message includes
            ``consent_url`` so the caller can redirect the user.

    Example::

        from cred_autogen import cred_delegate_tool

        tool = cred_delegate_tool(
            agent_token=os.environ["CRED_AGENT_TOKEN"],
            user_id="user_123",
            app_client_id="my_app",
        )
        # Register with an AutoGen agent
    """
    if token_format not in {"raw", "handle"}:
        raise ValueError("token_format must be 'raw' or 'handle'")

    cred = Cred(agent_token=agent_token, base_url=base_url)

    def delegate(
        service: Annotated[str, "Service slug to get a token for (e.g. 'google', 'github', 'google-calendar')."],
        scopes: Annotated[list[str], "OAuth scopes to request (e.g. ['calendar.readonly']). Pass an empty list to use all consented scopes."],
    ) -> str:
        """Get a delegated OAuth access token for a third-party service on behalf of the current user.

        Returns JSON with access_token, token_type, expires_in, service, scopes,
        and delegation_id. Raises an error with consent_url if the user hasn't
        connected the service.
        """
        resolved_scopes: Optional[list[str]] = scopes if scopes else None

        if token_format == "handle":
            result = cred.delegate_handle(
                service=service,
                user_id=user_id,
                app_client_id=app_client_id,
                scopes=resolved_scopes,
            )
            return json.dumps({
                "token_type": result.token_type,
                "expires_in": result.expires_in,
                "service": result.service,
                "user_id": result.user_id,
                "scopes": result.scopes,
                "delegation_id": result.delegation_id,
                "note": "Pass delegation_id to cred_use to make authenticated API calls.",
            })

        result = cred.delegate(
            service=service,
            user_id=user_id,
            app_client_id=app_client_id,
            scopes=resolved_scopes,
        )
        return json.dumps({
            "access_token": result.access_token,
            "token_type": result.token_type,
            "expires_in": result.expires_in,
            "service": result.service,
            "scopes": result.scopes,
            "delegation_id": result.delegation_id,
        })

    description = (
        ("Get a brokered OAuth delegation handle for a third-party service "
         if token_format == "handle"
         else "Get a delegated OAuth access token for a third-party service ") +
        "on behalf of the current user. " +
        ("Returns token_type, expires_in, service, user_id, scopes, and delegation_id. Pass delegation_id to cred_use. "
         if token_format == "handle"
         else "Returns access_token, token_type, expires_in, service, and scopes. ") +
        "Raises an error with consent_url if the user hasn't connected the service."
    )
    return FunctionTool(delegate, name="cred_delegate", description=description)


def cred_use_tool(
    agent_token: str,
    base_url: str,
) -> FunctionTool:
    """Create an AutoGen FunctionTool for brokered upstream API calls."""
    cred = Cred(agent_token=agent_token, base_url=base_url)

    def use(
        delegation_id: Annotated[str, "Brokered delegation handle returned by cred_delegate."],
        url: Annotated[str, "Full HTTPS API URL to call."],
        method: Annotated[str, "HTTP method: GET, POST, PUT, PATCH, or DELETE."],
        body: Annotated[Optional[dict[str, Any]], "Optional JSON request body."] = None,
        extra_headers: Annotated[Optional[dict[str, str]], "Optional service-specific headers."] = None,
    ) -> str:
        """Make an authenticated API call through Cred using a brokered delegation handle."""
        result = cred.use(
            delegation_id=delegation_id,
            url=url,
            method=method,
            body=body,
            extra_headers=extra_headers,
        )
        return json.dumps({
            "status": result.status,
            "ok": result.ok,
            "content_type": result.content_type,
            "body": result.body,
            "truncated": result.truncated,
            "truncated_at": result.truncated_at,
        })

    return FunctionTool(use, name="cred_use", description=(
        "Make an authenticated API call through Cred using a brokered delegation handle. "
        "The provider access token stays on the Cred server."
    ))

File: packages/integrations/vercel-ai/src/tool.ts
/**
 * Cred Vercel AI SDK integration.
 *
 * Returns Vercel AI SDK tools that agents can call for delegated OAuth access.
 * The tool schema matches the Cred MCP tool spec:
 *
 *   name:   cred_delegate
 *   params: { service: string, scopes: string[] }
 *
 * userId and appClientId are pre-configured at factory time and are not
 * agent-controlled inputs.
 */

import { tool, type Tool } from 'ai';
import { z } from 'zod';
import { Cred } from '@credninja/sdk';

interface CredDelegateToolInput {
  service: string;
  scopes: string[];
}


export interface CredDelegateToolOptions {
  /** Agent token issued by Cred (starts with cred_at_) */
  agentToken: string;
  /** The user whose credentials to delegate */
  userId: string;
  /** The Cred app client ID to delegate for */
  appClientId: string;
  /** Your Cred server URL (e.g. https://cred.example.com or http://localhost:3456) */
  baseUrl: string;
  /**
   * Token delivery mode. Defaults to `handle` (brokered): the provider access
   * token stays on the Cred server and only a delegation handle is returned to
   * the model — use it with `credUseTool`. Set to `raw` only if you explicitly
   * need the access token returned to the model; this exposes a live OAuth
   * token to the LLM context and defeats the brokered security model.
   */
  tokenFormat?: 'raw' | 'handle';
}

export interface CredUseToolOptions {
  /** Agent token issued by Cred (starts with cred_at_) */
  agentToken: string;
  /** Your Cred server URL (e.g. https://cred.example.com or http://localhost:3456) */
  baseUrl: string;
}

interface CredUseToolInput {
  delegationId: string;
  url: string;
  method: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
  body?: Record<string, unknown>;
  extraHeaders?: Record<string, string>;
}

/**
 * Create a Vercel AI SDK tool for credential delegation.
 *
 * The returned tool is pre-configured with the user's identity and app context.
 * At runtime, the agent supplies `service` and `scopes`.
 *
 * @example
 * ```typescript
 * import { credDelegateTool } from '@credninja/ai';
 * import { generateText } from 'ai';
 *
 * const tool = credDelegateTool({
 *   agentToken: process.env.CRED_AGENT_TOKEN!,
 *   baseUrl: process.env.CRED_BASE_URL!,
 *   userId: 'user_123',
 *   appClientId: 'my_app',
 * });
 *
 * const result = await generateText({
 *   model: yourModel,
 *   tools: { cred_delegate: tool },
 *   prompt: 'Get my Google Calendar events',
 * });
 * ```
 */
export function credDelegateTool(options: CredDelegateToolOptions): Tool {
  const { agentToken, userId, appClientId, baseUrl, tokenFormat = 'handle' } = options;
  const inputSchema = z.object({
    service: z
      .string()
      .describe("Service slug to get a token for (e.g. 'google', 'github', 'google-calendar')."),
    scopes: z
      .array(z.string())
      .describe(
        "OAuth scopes to request (e.g. ['calendar.readonly']). " +
          'Pass an empty array to use all consented scopes.',
      ),
  }) as z.ZodType<CredDelegateToolInput>;

  const cred = new Cred({
    agentToken,
    baseUrl,
  });
  let cachedResult:
    | {
        accessToken?: string;
        tokenType: string;
        expiresIn: number;
        service: string;
        scopes: string[];
        delegationId: string;
        userId?: string;
        expiresAtMs: number;
        cacheKey: string;
      }
    | null = null;

  const toolDefinition = {
    description:
      (tokenFormat === 'handle'
        ? 'Get a brokered OAuth delegation handle for a third-party service '
        : 'Get a delegated OAuth access token for a third-party service ') +
      'on behalf of the current user. ' +
      (tokenFormat === 'handle'
        ? 'Returns token_type, expires_in, service, scopes, user_id, and delegation_id. Pass delegation_id to cred_use. '
        : 'Returns access_token, token_type, expires_in, service, scopes, and delegation_id. ') +
      'Raises an error with consent_url if the user has not connected the service.',
    inputSchema,
    execute: async ({ service, scopes }: CredDelegateToolInput) => {
      const resolvedScopes = scopes.length > 0 ? scopes : undefined;
      const cacheKey = JSON.stringify({ service, scopes: resolvedScopes ?? [] });
      const refreshCutoffMs = Date.now() + 60_000;

      if (cachedResult && cachedResult.cacheKey === cacheKey && cachedResult.expiresAtMs > refreshCutoffMs) {
        return {
          ...(cachedResult.accessToken ? { accessToken: cachedResult.accessToken } : {}),
          tokenType: cachedResult.tokenType,
          expiresIn: cachedResult.expiresIn,
          service: cachedResult.service,
          scopes: cachedResult.scopes,
          delegationId: cachedResult.delegationId,
          ...(cachedResult.userId ? { userId: cachedResult.userId } : {}),
          ...(tokenFormat === 'handle' ? { note: 'Pass delegationId to cred_use to make authenticated API calls.' } : {}),
        };
      }

      const result = tokenFormat === 'handle'
        ? await cred.delegateHandle({
            service,
            userId,
            appClientId,
            scopes: resolvedScopes,
          })
        : await cred.delegate({
            service,
            userId,
            appClientId,
            scopes: resolvedScopes,
          });

      const expiresAtMs = result.expiresAt instanceof Date
        ? result.expiresAt.getTime()
        : Date.now() + result.expiresIn * 1000;
      cachedResult = {
        ...('accessToken' in result ? { accessToken: result.accessToken } : {}),
        tokenType: result.tokenType,
        expiresIn: result.expiresIn,
        service: result.service,
        scopes: result.scopes,
        delegationId: result.delegationId,
        ...('userId' in result ? { userId: result.userId } : {}),
        expiresAtMs,
        cacheKey,
      };

      return {
        ...('accessToken' in result ? { accessToken: result.accessToken } : {}),
        tokenType: result.tokenType,
        expiresIn: result.expiresIn,
        service: result.service,
        scopes: result.scopes,
        delegationId: result.delegationId,
        ...('userId' in result ? { userId: result.userId } : {}),
        ...(tokenFormat === 'handle' ? { note: 'Pass delegationId to cred_use to make authenticated API calls.' } : {}),
      };
    },
  };

  return tool(toolDefinition as any);
}

/**
 * Create a Vercel AI SDK tool that brokers an upstream API request through Cred.
 *
 * Use this with `credDelegateTool({ tokenFormat: 'handle', ... })`.
 */
export function credUseTool(options: CredUseToolOptions): Tool {
  const { agentToken, baseUrl } = options;
  const cred = new Cred({
    agentToken,
    baseUrl,
  });

  const inputSchema = z.object({
    delegationId: z.string().describe('Delegation handle returned by cred_delegate.'),
    url: z.string().url().describe('Full HTTPS API URL to call. The Cred server enforces service allowlists.'),
    method: z.enum(['GET', 'POST', 'PUT', 'PATCH', 'DELETE']).describe('HTTP method.'),
    body: z.record(z.string(), z.unknown()).optional().describe('JSON request body for POST, PUT, or PATCH.'),
    extraHeaders: z.record(z.string(), z.string()).optional().describe('Optional service-specific headers.'),
  }) as z.ZodType<CredUseToolInput>;

  return tool({
    description:
      'Make an authenticated API call through Cred using a brokered delegation handle. ' +
      'The provider access token stays on the Cred server.',
    inputSchema,
    execute: async ({ delegationId, url, method, body, extraHeaders }: CredUseToolInput) => {
      return cred.use({
        delegationId,
        url,
        method,
        body,
        extraHeaders,
      });
    },
  } as any);
}
